### PR TITLE
Issue 5554: Test Consumption based Stream Retention with Single Subscriber

### LINKED ISSUE #5554: Test Consumption based Stream Retention with Single Subscriber
**Problem description**
_This test verifies that:_
1. Reads of Non-Subscriber Reader Groups do not impact Stream Truncation.
2. Reads of Subscriber Reader Groups impact Stream truncation only when a truncation stream-cut is published to Controller.
3. When truncation stream-cuts are **not** published, stream truncation happens based on min/max limits set in the retention policy.
4. When truncation stream-cuts **are** published by a Subscriber Reader Group, Stream truncation happens based on these stream-cuts.

_Test Steps:_
1. Create a Stream with SPACE based retention policy with min/max limits.
2. Set Controller retention cycle interval to be 1 minute.
3. Create a Non-Subscriber Reader Groups to read data from this Stream. 
4. Create a writer that writes more data to the Stream than the max limit.
5. Non-Subscriber RG starts reading data from the Stream.
6. Verify - After 1 minute, Stream gets truncated at max limit.
7. Create a new Subscriber Reader Group to read from the same Stream and set its RetentionType=MANUAL_RELEASE_AT_USER_STREAMCUT and enable automatic checkpointing.
8. Write more then max_limit data into the Stream and wait for 1 minute.
9. Stream should get truncated at the Stream-Cut that satisfies the max_limit for the Stream's Retention Policy, since we haven't yet published any truncation Stream-Cut.
10. On the Subscriber Reader Group create a new checkpoint using initiateCheckpoint() API
11. Publish this checkpoint as the truncation point for the Stream using updateRetentionStreamCut() API on this ReaderGroup
12. Verify that when the retention cycle runs, the Stream is truncated at this point.
13. Change Subscriber ReaderGroup configuration to RetentionType=AUTOMATIC_RELEASE_AT_LAST_CHECKPOINT.
14. Verify that when the retention cycle runs, the Stream does get truncated at the last-checkpoint stream-cut of the Subscriber ReaderGroup. Note that this will happen only if the published truncation Stream-Cut falls between `min` and `max` space limits set for the Stream. So Stream should be written to accordingly.

**Problem location**
System Tests

**Suggestions for an improvement**
Create a new Test Class ConsumptionBasedRetentionTest

### PR DESCRIPTION
**Pull request by anirudhkovuru**
**Change log description**  
Added a new system test to check Consumption based Stream Retention through a single subscriber.

**Purpose of the change**  
Fixes #5554.

**What the code does**  
1. A test stream with a single segment is created with a space based retention policy where the minimum to retain is 30 bytes and the maximum to retain is 300 bytes.
2. An event (e1) of size 30 bytes is written into the stream.
3. A single subscriber, which is subscribed to the test stream, is created. The subscriber reads the event and updates the retention stream-cut with the one after this event (e1) to the controller.
4. Since the size retained by the provided stream-cut would be 0 bytes, truncation does not take place as it is not within the min and max bounds.
5. Two more events (e2 and e3) of size 30 bytes each are written into the stream.
6. Since the size retained by the provided stream-cut would be 60 bytes now, truncation should take place at the provided stream-cut as it is within the min and max bounds.
7. The subscriber reads another event and updates its retention stream-cut to the one after the read event (e2).
8. Since the size retained by the provided stream-cut would be 30 bytes now, truncation should take place at the provided stream-cut as it is within the min and max bounds.

**How to verify it**  
All tests and system tests must pass.

### CODE REVIEW

File: test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java
Comment (andreipaduroiu) [inline test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java:158]: This makes it a total of 10 minutes.

File: test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java
Comment (andreipaduroiu) [inline test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java:174]: 15 minutes.

**Review by andreipaduroiu — CHANGES_REQUESTED** [review]
Please rewrite your test and/or introduce any necessary config changes into the code to avoid any kind of unnecessary sleeps.

File: test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java
Comment (anirudhkovuru) [reply test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java]: I've removed the 5 min timer.

**Review by andreipaduroiu — COMMENTED** [review]
This looks better. How long does this system test take to run? 

@shiveshr / @pbelgundi  Please take a look at this PR. I cannot remove my Changes Requested without approving it (which will unblock it for merging), so I will await one of your approvals before I approve it too.

**Comment by RaulGracia** [comment]
@anirudhkovuru you will need to change the header for this PR. 
@shiveshr @pbelgundi can you review this one? Looks like we can give a go to it soon.

File: test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java
Comment (shiveshr) [inline test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java:159]: this assert will effectively return immediately and we do not give it time to actually truncate. so we dont really verify anything at this step. 

verifying truncation doesnt happen is hard without relying on time.. 
so we might as well remove this check here.

Comment (shiveshr) [reply test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java:159]: also, anymatch offset=0 is not ideal either if tehre are more than one segments and all data got written to only one of them.. then only that one would have truncated.. 

your intent here should have been `allmatch` offset = 0

Comment (shiveshr) [reply test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java:159]: i know your configuration has only 1 segment.. but generally allmatch in this case is better representative of the logic in the code.

**Review by shiveshr — APPROVED** [review]
One minor comment

File: test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java
Comment (RaulGracia) [inline test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java:142]: `String.valueOf(1)` this is not needed, just use `1` and the concatenation will make it String automatically.

File: test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java
Comment (RaulGracia) [inline test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java:147]: Why don't you check that what you are reading is the right data? It could be even a checkpoint, right?

File: test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java
Comment (RaulGracia) [inline test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java:150]: What is this used for?

File: test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java
Comment (RaulGracia) [inline test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java:171]: Again, why do you need this?

**Review by RaulGracia — CHANGES_REQUESTED** [review]
@anirudhkovuru are you still pursuing this one?

**Comment by anirudhkovuru** [comment]
The following system test run for this branch passed: https://asd-nautilus-jenkins.isus.emc.com/job/pravega-branch-builder/1296/

File: test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java
Comment (anirudhkovuru) [reply test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java]: changed it to `1`

File: test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java
Comment (anirudhkovuru) [reply test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java]: There are checks to make sure the data read is what is expected

File: test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java
Comment (anirudhkovuru) [reply test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java:161]: In order to trigger the silent checkpoint generated during `generateStreamCuts` to complete we read another event after the `ReaderGroupState` updates itself.

File: test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java
Comment (anirudhkovuru) [reply test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java:190]: Explained above.

**Review by RaulGracia — COMMENTED** [review]
I have left some comments, but I'm not blocking the PR. Also, it would be interesting for this test to be a bit more realistic. Wondering if you could add another test method (in the same class) that could write/consume more data and with parallel threads. The current test is super small and unit-like, and for system tests we can try more realistic loads and validate the semantics from an application perspective.

**Review by RaulGracia — COMMENTED** [review]
Also, it would be interesting for this test to be a bit more realistic. Wondering if you could add another test method (in the same class) that could write/consume more data and with parallel threads. The current test is super small and unit-like, and for system tests we can try more realistic loads and validate the semantics from an application perspective.

File: test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java
Comment (anirudhkovuru) [reply test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java]: Using self describing names

File: test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java
Comment (anirudhkovuru) [reply test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java:178]: The retention cycle takes place periodically every 2 min. By waiting for 5 min we allow the retention cycle to run twice, makes sure that the truncation we wish to see is guaranteed to happen.

File: test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java
Comment (anirudhkovuru) [reply test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java]: Used a variable

File: test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java
Comment (anirudhkovuru) [reply test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java]: Fixed that.

File: test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java
Comment (pbelgundi) [inline test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java:161]: I don't like the idea of adding sleeps to tests as it increases possibility of making tests flaky. Why do we need to wait for reader group state update here?

File: test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java
Comment (pbelgundi) [reply test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java:178]: The frequency with which Controller executes the retention cycle is a configurable parameter that can be specified in ControllerServiceConfigImpl. See: https://github.com/pravega/pravega/blob/master/controller/src/main/java/io/pravega/controller/server/impl/ControllerServiceConfigImpl.java#L106
When not explicitly set this value defaults to Config.MINIMUM_RETENTION_FREQUENCY_IN_MINUTES which is 1 min...
So there should not be a need to wait for 5 mins...

**Review by pbelgundi — CHANGES_REQUESTED** [review]

File: test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java
Comment (pbelgundi) [reply test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java:178]: Another option can be to AssertEventuallyEquals() that head of the Stream is equal to the StreamCut at which we expect truncation to happen. There should be no need to specify any time interval...

**Comment by pbelgundi** [comment]
> t could write/consume more data and with parallel threads. The current test is super small and unit-like, and for system

This is supposed to be a basic CBR test that covers 1 most basic CBR scenario. More tests with multiple reader groups and more complex scenarios would be added later. See: https://github.com/pravega/pravega/issues/5553

File: test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java
Comment (anirudhkovuru) [reply test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java:161]: Calling `generateStreamCuts` leads to the creation of a silent checkpoint. In order for the ReaderGroup to process the checkpoint we need to wait and allow the ReaderGroupState to update and fetch the checkpoint so that it can be processed. This is present in the form of a timer (`checkpointTimer`) in the `ReaderGroupStateManager`.

File: test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java
Comment (anirudhkovuru) [reply test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java]: Changed the name accordingly.

**Comment by anirudhkovuru** [comment]
Regarding the 5 minute wait for periodic retention to kick in. This was how it was implemented in other system tests. So the aspect of dealing with that wait can be dealt with in different issue where we can address all such cases at once. The 5 second reader group state update timeout can be dealt with similarly.

**Comment by anirudhkovuru** [comment]
A successful system test run: https://asd-nautilus-jenkins.isus.emc.com/job/pravega-branch-builder/1499/

File: test/system/src/test/java/io/pravega/test/system/SingleSubscriberUpdateRetentionStreamCutTest.java
Comment (RaulGracia) [inline test/system/src/test/java/io/pravega/test/system/SingleSubscriberUpdateRetentionStreamCutTest.java:82]: Why do you need 2 executors?

File: test/system/src/test/java/io/pravega/test/system/SingleSubscriberUpdateRetentionStreamCutTest.java
Comment (RaulGracia) [inline test/system/src/test/java/io/pravega/test/system/SingleSubscriberUpdateRetentionStreamCutTest.java:123]: You have actually missed to shut down the second executor.

File: test/system/src/test/java/io/pravega/test/system/SingleSubscriberUpdateRetentionStreamCutTest.java
Comment (RaulGracia) [inline test/system/src/test/java/io/pravega/test/system/SingleSubscriberUpdateRetentionStreamCutTest.java:162]: Let's do not add more sleeps if possible. If the update takes 6 seconds this test will fail, and if it takes 1 second we are wasting 4 seconds. Please, use `assertEventuallyEquals` if possible, as it would check the state of the reader group until the condition you want is true.

File: test/system/src/test/java/io/pravega/test/system/SingleSubscriberUpdateRetentionStreamCutTest.java
Comment (RaulGracia) [inline test/system/src/test/java/io/pravega/test/system/SingleSubscriberUpdateRetentionStreamCutTest.java:179]: I think that this is checking every few milliseconds by default. Maybe you can pick a less aggressive evaluation interval, given that this is going to take at least a minute.

File: test/system/src/test/java/io/pravega/test/system/SingleSubscriberUpdateRetentionStreamCutTest.java
Comment (RaulGracia) [inline test/system/src/test/java/io/pravega/test/system/SingleSubscriberUpdateRetentionStreamCutTest.java:194]: Same here about sleeps.

**Review by RaulGracia — CHANGES_REQUESTED** [review]

File: test/system/src/test/java/io/pravega/test/system/SingleSubscriberUpdateRetentionStreamCutTest.java
Comment (anirudhkovuru) [reply test/system/src/test/java/io/pravega/test/system/SingleSubscriberUpdateRetentionStreamCutTest.java:82]: One executor handles the controller, the other executor is for the client to process the silent checkpoint generated as part of the `generateStreamCut` process. The same has been done in the `StreamCutsTest` system test.

File: test/system/src/test/java/io/pravega/test/system/SingleSubscriberUpdateRetentionStreamCutTest.java
Comment (anirudhkovuru) [reply test/system/src/test/java/io/pravega/test/system/SingleSubscriberUpdateRetentionStreamCutTest.java:123]: I added the shutdown

File: test/system/src/test/java/io/pravega/test/system/SingleSubscriberUpdateRetentionStreamCutTest.java
Comment (anirudhkovuru) [reply test/system/src/test/java/io/pravega/test/system/SingleSubscriberUpdateRetentionStreamCutTest.java:163]: This is being used in other tests too so we can try dealing with this as a separate issue.

File: test/system/src/test/java/io/pravega/test/system/SingleSubscriberUpdateRetentionStreamCutTest.java
Comment (anirudhkovuru) [reply test/system/src/test/java/io/pravega/test/system/SingleSubscriberUpdateRetentionStreamCutTest.java:180]: Similarly for this timeout.

File: test/system/src/test/java/io/pravega/test/system/SingleSubscriberUpdateRetentionStreamCutTest.java
Comment (pbelgundi) [inline test/system/src/test/java/io/pravega/test/system/SingleSubscriberUpdateRetentionStreamCutTest.java:128]: Can we call this singleSubscriberCBR ?

Comment (anirudhkovuru) [reply test/system/src/test/java/io/pravega/test/system/SingleSubscriberUpdateRetentionStreamCutTest.java]: I've changed this name.

File: test/system/src/test/java/io/pravega/test/system/SingleSubscriberUpdateRetentionStreamCutTest.java
Comment (anirudhkovuru) [reply test/system/src/test/java/io/pravega/test/system/SingleSubscriberUpdateRetentionStreamCutTest.java]: I've changed the check interval to 1 min.

Comment (RaulGracia) [reply test/system/src/test/java/io/pravega/test/system/SingleSubscriberUpdateRetentionStreamCutTest.java]: Well, if this is going to take 1 minute, what can happen is that you evaluate one second before the condition evaluates to true, and then the test waits for 1 minute doing nothing. Evaluating every 1 or 5 seconds should be fine.

Comment (anirudhkovuru) [reply test/system/src/test/java/io/pravega/test/system/SingleSubscriberUpdateRetentionStreamCutTest.java]: I've changed it to 1 sec

**Review by RaulGracia — APPROVED** [review]

**Review by andreipaduroiu — APPROVED** [review]

**Comment by anirudhkovuru** [comment]
The system test run is green for this PR. The build is 1561 on pravega-branch-builder and 4083 on test-pravega-continuous.

**Comment by RaulGracia** [comment]
@pbelgundi can you please have another look? This seems close to get merged.

**Review by pbelgundi — APPROVED** [review]
LGTM

**Pull request merged**
